CPOL-42 Allows pagination

josejulio opened the pull request:
~~I'm adding `lombok`, it needs IDE configuration to function properly:~~

~~Eclipse: https://projectlombok.org/setup/eclipse
Intelli: https://projectlombok.org/setup/intellij~~

Removed it

pilhuhn commented on src/main/java/com/redhat/cloud/custompolicies/app/rest/utils/PagingUtils.java:
```diff
@@ -0,0 +1,68 @@
+        try {
+            pageBuilder.page(Integer.parseInt(queryParams.getFirst("page")));
+        } catch (NumberFormatException nfe) {
+            // Ignore the exception, the builder will use the default value
```
Throwing and catching exceptions  is expensive. Better check if a parameter is present

burmanm replied:
In the context of this application, Exception is cheap (we're talking about hundreds of millions of ops per second). However, there's no replacement for Exception here - the parseInt can throw Exception if the user sends "page=second".

The question is rather - should we really use the default or return this as error to the user?

josejulio replied:
> However, there's no replacement for Exception here - the parseInt can throw Exception if the user sends "page=second".

That's the use there, right.

> The question is rather - should we really use the default or return this as error to the user?

Good point.
I'll better throw the exception to let the user know about this error.

pilhuhn commented on src/main/java/com/redhat/cloud/custompolicies/app/rest/utils/PagingUtils.java:
```diff
@@ -0,0 +1,68 @@
+        try {
+            pageBuilder.itemsPerPage(Integer.parseInt(queryParams.getFirst("pageSize")));
+        } catch (NumberFormatException nfe) {
+            // Ignore the exception, the builder will use the default value
```
Same as above

pilhuhn commented on pom.xml:
```diff
@@ -119,6 +119,12 @@
     </dependency>
+    <dependency>
+      <groupId>org.projectlombok</groupId>
+      <artifactId>lombok</artifactId>
```
I don't think this is a good idea. This pulls in a large lib for little gain.

josejulio replied:
I was more interested on the `Builder`, looks more DRY and is configurable as simple annotations. 
But I agree that is a little gain for now. I'll remove the lib.

josejulio commented:
Updated the code: 
 - no more `lombok`, added the builder and getters.
 - NFE are reported to the client, but still using defaults if page or pagePerItem are not passed.

pilhuhn commented on src/main/java/com/redhat/cloud/custompolicies/app/rest/utils/PagingUtils.java:
```diff
@@ -0,0 +1,85 @@
+            builder = Response.ok(page);
+            EntityTag etag = new EntityTag(String.valueOf(page.hashCode()));
+            builder.header("ETag",etag);
+            builder.header("TotalCount", Long.toString(page.getTotalCount()));
```
I don't know for sure, but  I think custom headers should have a `X-` prefix. Doesn't really harm leaving it as it I guess

josejulio replied:
I was reading that the `X-` prefix has been deprecated: https://tools.ietf.org/html/rfc6648
